feat(cmp) CMP-53 : Modify Aiven Http KConnect : adding new authorizat…

z24cbreu opened the pull request:
…ion type

- add configuration apikey
- call api key url with key and secret configuration to get access_token
- request builder for http apikey sender
- adding https protocols in properties

z24cbreu committed:
feat(cmp) CMP-53 : Modify Aiven Http KConnect : adding new authorization type
- add configuration apikey
- call api key url with key and secret configuration to get access_token
- request builder for http apikey sender
- adding https protocols in properties

aprodhomme-dkt commented on build.gradle:
```diff
@@ -25,7 +25,7 @@ plugins {
 
     // https://docs.gradle.org/current/userguide/checkstyle_plugin.html
-    id 'checkstyle'
+    //id 'checkstyle'
```
why have you commented this line?

aprodhomme-dkt commented on build.gradle:
```diff
@@ -164,9 +164,9 @@ dependencies {
-checkstyle {
-    toolVersion '8.25'
-}
+//checkstyle {
```
+1 on the commented lines?

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:
```diff
@@ -314,6 +319,60 @@ public String toString() {
+                new ConfigDef.NonEmptyStringWithoutControlChars() {
+                    @Override
+                    public String toString() {
+                        return "ApiKey scope";
```
Don't you want to specify which field is wrong when the validator returns false? eg: "ApiKey scope - URL"

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:
```diff
@@ -314,6 +319,60 @@ public String toString() {
+                new ConfigDef.NonEmptyStringWithoutControlChars() {
+                    @Override
+                    public String toString() {
+                        return "ApiKey scope";
```
"ApiKey Scope - Key"

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:
```diff
@@ -314,6 +319,60 @@ public String toString() {
+                new ConfigDef.NonEmptyStringWithoutControlChars() {
+                    @Override
+                    public String toString() {
+                        return "ApiKey scope";
```
"ApiKey Scope - Secret"

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:
```diff
@@ -314,6 +319,60 @@ public String toString() {
+                    }
+                },
+                ConfigDef.Importance.LOW,
+                "The key value of the body sent to REST API returning the access token",
```
I think the documentation is not accurate. IMO it would be something like "The authentication key sent to the REST API returning the access token"

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:
```diff
@@ -314,6 +319,60 @@ public String toString() {
+                    }
+                },
+                ConfigDef.Importance.LOW,
+                "The secret value of the body sent to REST API returning the access token",
```
same remark, I would go for something like "The authentication secret sent to the REST API returning the access token"

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyAccessTokenHttpRequestBuilder.java:
```diff
@@ -0,0 +1,76 @@
+            return accessTokenRequestBuilder
+                    .POST(HttpRequest.BodyPublishers.ofString(requestBody));
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
```
I have some doubts on how this RuntimeException will be handled. This part needs some rework to property handle the exception.

aprodhomme-dkt replied:
You still need to handle this exception. Why not add the throwing of IOException on the build method?
eg:
```
interface HttpRequestBuilder {
...

    HttpRequest.Builder build(final HttpSinkConfig config) throws IOException;
```
You just need to add this exception on the interface and add it on implementations, that's all. The IOException is already handled by the calling method `loadOrRefreshAccessToken`.

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSender.java:
```diff
@@ -0,0 +1,77 @@
+package io.aiven.kafka.connect.http.sender;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.ObjectCodec;
```
unused import

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSender.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    protected HttpResponse<String> sendWithRetries(HttpRequest.Builder requestBuilder, HttpResponseHandler httpResponseHandler) {
```
this method should be in the super class, it's duplicated from OAuth2HttpSender

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSender.java:
```diff
@@ -0,0 +1,77 @@
+
+    private HttpRequest.Builder loadOrRefreshAccessToken(final HttpRequest.Builder requestBuilder) {
+        if (accessTokenAuthHeader == null) {
+            logger.info("Configure ApiKey for URI: {} and Key: {}",
```
do we need to have an INFO level or DEBUG might enough?

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSender.java:
```diff
@@ -0,0 +1,77 @@
+                                HttpResponseHandler.ON_HTTP_ERROR_RESPONSE_HANDLER
+                        );
+                accessTokenAuthHeader = buildAccessTokenAuthHeader(response.body());
+            } catch (final IOException e) {
```
we would expect the JsonFormattingException to be catched here

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSender.java:
```diff
@@ -0,0 +1,77 @@
+        return requestBuilder.setHeader(HttpRequestBuilder.HEADER_AUTHORIZATION, accessTokenAuthHeader);
+    }
+
+    private String buildAccessTokenAuthHeader(final String responseBody) throws JsonProcessingException {
```
Your method is doing the exact same thing as the one from OAuth2HttpSender.

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSender.java:
```diff
@@ -0,0 +1,77 @@
+import java.net.http.HttpResponse;
+import java.util.Map;
+
+public class ApiKeyHttpSender extends HttpSender {
```
To my point of view, you should refactor the OAuth2HttpSender to pass a factory that would create the proper HttpRequestBuilder from HttpSender.createHttpSender

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/HttpSender.java:
```diff
@@ -108,6 +108,8 @@ public static HttpSender createHttpSender(final HttpSinkConfig config) {
             case OAUTH2:
                 return new OAuth2HttpSender(config);
+            case APIKEY:
+                return new ApiKeyHttpSender(config);
```
you could give the right Builder here and then use it in the OAuth2HttpSender

aprodhomme-dkt commented on src/test/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSenderTest.java:
```diff
@@ -0,0 +1,299 @@
+
+        when(accessTokenResponse.statusCode()).thenReturn(200);
+        when(accessTokenResponse.body()).thenReturn(objectMapper.writeValueAsString(accessTokenJson));
+        when(mockedHttpClient.<String>send(requestCaptor.capture(), any())).thenReturn(accessTokenResponse);
```
this test is not very useful, you don't check that the you are calling with the ApiKey here. Using any() will answer a mocked accessToken to any request.

aprodhomme-dkt commented on src/test/java/io/aiven/kafka/connect/http/sender/ApiKeyHttpSenderTest.java:
```diff
@@ -0,0 +1,299 @@
+    final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Test
+    void buildAccessTokenAuthHeaderForDefaultSettings(@Mock final HttpResponse<String> accessTokenResponse)
```
this method is the only one you should keep because it tests the token retrieval from an api key. The other test methods are duplicates from the OAuth2HttpSenderTest class

z24cbreu committed:
feat(cmp) CMP-53 : PR review
- remove ApiKeyHttpSender.java and ApiKeyHttpSenderTest.java
- use of OAuth2HttpSender instead with a custom AccessTokenRequestBuilder

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:
```diff
@@ -523,6 +521,26 @@ OAUTH2_ACCESS_TOKEN_URL_CONFIG, getString(OAUTH2_ACCESS_TOKEN_URL_CONFIG),
                                     + " = " + AuthorizationType.OAUTH2);
                 }
                 break;
+            case APIKEY:
```
everything is now duplicated with OAUTH2. This could be simplified with:
```
        var authorizationType = authorizationType();
        switch (authorizationType) {
            case STATIC:
               ...
            case OAUTH2:
            case APIKEY:
                if(oauth2AccessTokenUri() == null) {
                    throw new ConfigException(
                            OAUTH2_ACCESS_TOKEN_URL_CONFIG, getString(OAUTH2_ACCESS_TOKEN_URL_CONFIG),
                            "Must be present when " + HTTP_HEADERS_CONTENT_TYPE_CONFIG
                                    + " = " + authorizationType);
                }
...
```

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/ApiKeyAccessTokenHttpRequestBuilder.java:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    static class AccessTokenRequest {
```
this inner class should be private and immutable by setting all of this field to final and get rid of the setters

```
final String type;
        final String key;
        final String secret;

        public AccessTokenRequest(String type, String key, String secret) {
            this.type = type;
            this.key = key;
            this.secret = secret;
        }
```

aprodhomme-dkt commented on src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java:
```diff
@@ -64,14 +66,14 @@ protected HttpResponse<String> sendWithRetries(final HttpRequest.Builder request
     }
 
-    private HttpRequest.Builder loadOrRefreshAccessToken(final HttpRequest.Builder requestBuilder) {
+    private HttpRequest.Builder loadOrRefreshAccessToken(final HttpRequest.Builder requestBuilder, HttpRequestBuilder accessTokenRequestBuilder ) {
```
you don't need to pass the builder in the parameters as you already have it your class instance

aprodhomme-dkt reviewed (changes requested):
You have to test you builder ApiKeyAccessTokenHttpRequestBuilder with unit tests with at least 2 tests:
* a passing case
* a case with a JsonProcessingException

The README is not up to date. The ApiKey authorization is not listed.
Add an integration test to test the ApiKey configuration into IntegrationTest.java

z24cbreu committed:
feat(cmp) CMP-53 : PR review
- Use of a custom builder for config api-key
- cleaning config

aprodhomme-dkt committed:
feat(cmp) cmp-53: final review

- apply checkstyle fixes
- add unit tests
- add specific interface for oauth2 request builders
